File: ActFinal/UnionFind.cpp
#include <iostream>
#include <vector>
using namespace std;

class DisjointSet {
private:
    vector<int> parent;
    vector<int> rank;

public:
    // Inicialización: cada nodo es su propia raíz
    DisjointSet(int n) {
        parent.resize(n);
        rank.resize(n, 0);
        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }
    }

    // Find con compresión de caminos (recursivo)
    int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]); // Compresión
        }
        return parent[x];
    }

    // Unión por rango
    void unionSets(int a, int b) {
        int rootA = find(a);
        int rootB = find(b);

        if (rootA != rootB) {
            if (rank[rootA] < rank[rootB]) {
                parent[rootA] = rootB;
            } else if (rank[rootA] > rank[rootB]) {
                parent[rootB] = rootA;
            } else {
                parent[rootB] = rootA;
                rank[rootA]++;
            }
            cout << ">>> Enlace establecido exitosamente entre " << a << " y " << b << "." << endl;
        } else {
            cout << ">>> Alerta: Los nodos " << a << " y " << b << " ya poseen conexion." << endl;
        }
    }

    // Consulta de conectividad
    bool isConnected(int a, int b) {
        return find(a) == find(b);
    }

    // Mostrar estructura interna
    void showStructure() {
        cout << "Padres: ";
        for (int i = 0; i < parent.size(); i++) {
            cout << parent[i] << " ";
        }
        cout << endl;
    }
};

int main() {
    int total_servidores = 6;
    DisjointSet red(total_servidores);

    cout << "=========================================" << endl;
    cout << "SISTEMA DE GESTION DE REDES DINAMICAS (C++)" << endl;
    cout << "=========================================\n" << endl;

    red.unionSets(0, 1);
    red.unionSets(1, 2);
    red.unionSets(3, 4);

    cout << "\n--- Verificacion de Canales de Comunicacion ---" << endl;
    cout << "Servidores 0 y 2: " << (red.isConnected(0, 2) ? "CONECTADOS" : "SIN CONEXION") << endl;
    cout << "Servidores 2 y 3: " << (red.isConnected(2, 3) ? "CONECTADOS" : "SIN CONEXION") << endl;

    cout << "\n--- Interconexion de Subredes (Enlazando 2 con 4) ---" << endl;
    red.unionSets(2, 4);

    cout << "\n--- Verificacion Final de Estado General ---" << endl;
    cout << "Servidores 0 y 3: " << (red.isConnected(0, 3) ? "CONECTADOS" : "SIN CONEXION") << endl;

    // Mostrar estructura interna
    cout << "\n--- Estructura interna (parent) ---" << endl;
    red.showStructure();

    return 0;
}

File: ActFinal/Union-Find/Union-Find/Union_Find.py
class DisjointSet:
    # Inicialización
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n

    # Find con compresión de caminos
    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])  # Compresión
        return self.parent[x]

    # Unión por rango
    def union_sets(self, a, b):
        root_a = self.find(a)
        root_b = self.find(b)

        if root_a != root_b:
            if self.rank[root_a] < self.rank[root_b]:
                self.parent[root_a] = root_b
            elif self.rank[root_a] > self.rank[root_b]:
                self.parent[root_b] = root_a
            else:
                self.parent[root_b] = root_a
                self.rank[root_a] += 1
            print(f">>> Enlace establecido exitosamente entre {a} y {b}.")
        else:
            print(f">>> Alerta: Los nodos {a} y {b} ya poseen conexion.")

    def is_connected(self, a, b):
        return self.find(a) == self.find(b)

    def show_structure(self):
        print("Padres:", self.parent)


if __name__ == "__main__":
    total_servidores = 6
    red = DisjointSet(total_servidores)

    print("=========================================")
    print("SISTEMA DE GESTION DE REDES DINAMICAS (PYTHON)")
    print("=========================================\n")

    red.union_sets(0, 1)
    red.union_sets(1, 2)
    red.union_sets(3, 4)

    print("\n--- Verificacion de Canales de Comunicacion ---")
    print(f"Servidores 0 y 2: {'CONECTADOS' if red.is_connected(0, 2) else 'SIN CONEXION'}")
    print(f"Servidores 2 y 3: {'CONECTADOS' if red.is_connected(2, 3) else 'SIN CONEXION'}")

    print("\n--- Interconexion de Subredes (Enlazando 2 con 4) ---")
    red.union_sets(2, 4)

    print("\n--- Verificacion Final de Estado General ---")
    print(f"Servidores 0 y 3: {'CONECTADOS' if red.is_connected(0, 3) else 'SIN CONEXION'}")

    print("\n--- Estructura interna (parent) ---")
    red.show_structure()
